app/gui2/src/workers/visualizationCompiler.ts

```diff
@@ -0,0 +1,298 @@
+        }
+        if (
+          isBuiltin ||
+          (isRelative && (path.endsWith('.svg') || path.endsWith('.ts') || path.endsWith('.vue')))
+        ) {
+          const specifiers = stmt.specifiers.map((s: any) => {
+            if (s.type === 'ImportDefaultSpecifier') {
+              return `default: ${s.local.name}`
+            } else {
+              if (s.imported.start === s.local.start) {
+                return s.imported.loc.identifierName
+              } else {
+                return `${s.imported.loc.identifierName}: ${s.local.loc.identifierName}`
+              }
+            }
+          })
+          s.overwrite(
+            stmt.start!,
+            stmt.end!,
+            `const { ${specifiers.join(
+              ', ',
+            )} } = await window.__visualizationModules[${JSON.stringify(path)}];`,
+          )
+          if (isBuiltin) {
+            // No further action is needed.
+          } else if (path.endsWith('.svg')) {
+            await importSvg(path)
+          } else if (path.endsWith('.ts')) {
+            await importTS(path)
+          } else if (path.endsWith('.vue')) {
+            await importVue(path)
+          }
+        }
+        break
+      }
+      case 'ExportDefaultDeclaration': {
+        if (id != null && (stmt.declaration as any)?.callee?.name === '_defineComponent') {
+          const firstProp = (stmt.declaration as any)?.arguments?.[0]?.properties?.[0]
+          if (firstProp?.start != null) {
+            s.appendLeft(firstProp.start, `__scopeId: ${JSON.stringify(`data-v-${id}`)}, `)
+          }
+        }
+        break
+      }
+    }
+  }
+  return s.toString()
+}
+
+async function compileVisualization(path: string, addStyle: (code: string) => void) {
+  const filename = path.match(/[^/\\]+$/)?.[0]!
+  const dir = path.replace(/[^/\\]+$/, '')
+  const text = await (await fetch(path)).text()
+  const id = generateId()
+  const parsed = parse(text, { filename })
+  for (const style of parsed.descriptor.styles) {
+    addStyle(
+      compileStyle({ filename, source: style.content, id, scoped: style.scoped ?? false }).code,
+    )
+  }
+  const scriptTs = compileScript(parsed.descriptor, {
+    id,
+    inlineTemplate: true,
+    sourceMap: false,
+  }).content
+  const scriptRaw = transform(scriptTs, {
+    disableESTransforms: true,
+    transforms: ['typescript'],
+  }).code
+  addImport(path, await rewriteImports(scriptRaw, dir, id))
+  return path
+}
+
+onmessage = async (
+  event: MessageEvent<
+    | { type: 'register-builtin-modules'; modules: string[] }
+    | { type: 'compile'; id: number; path: string }
+  >,
+) => {
+  switch (event.data.type) {
+    case 'register-builtin-modules': {
+      builtinModules = new Set(event.data.modules)
+      break
+    }
+    case 'compile': {
+      postMessage({
+        type: 'compilation-result',
+        id: event.data.id,
+        path: await compileVisualization(event.data.path, addStyle),
+      })
```

Comment:
Probably worth putting the `compileVisualization` into a `try` block, and doing a `postMessage` with an error when it throws one.

Reply:
added errors (+ refactored docs or viz compiler).